File: 1235-maximum-profit-in-job-scheduling/1235-maximum-profit-in-job-scheduling.java
class Solution {
    public int jobScheduling(int[] startTime, int[] endTime, int[] profit) {
         int N = startTime.length;
        List<Pair> pairJobSchedulingList = new ArrayList<>();
        for (int i = 0; i < N; ++i)
            pairJobSchedulingList.add(new Pair(i,startTime[i], endTime[i], profit[i]));
        pairJobSchedulingList.sort((a, b) -> a.endTime - b.endTime);
        TreeMap<Integer, Integer> orderProfit = new TreeMap<>();
        orderProfit.put(0, 0);
        for (Pair pair : pairJobSchedulingList) {
            int cur = orderProfit.floorEntry(pair.startTime).getValue() + pair.profit;
            if (cur > orderProfit.lastEntry().getValue())
                orderProfit.put(pair.endTime, cur);
        }
        return orderProfit.lastEntry().getValue();

    }
    private class Pair {
        int index;
        int startTime;
        int endTime;
        int profit;

public Pair(int index, int startTime, int endTime, int profit) {
            this.index = index;
            this.startTime = startTime;
            this.endTime = endTime;
            this.profit = profit;
        } 
    }
}

File: 1235-maximum-profit-in-job-scheduling/1235-maximum-profit-in-job-scheduling.py
from bisect import bisect, bisect_left
class Solution:
    def jobScheduling(self, startTime: List[int], endTime: List[int], profit: List[int]) -> int:
        def findMaxProfit(curr):
            print(curr)
            if curr >= n: return 0
            if memo[curr] != -1: return memo[curr]
            next_idx = bisect_left(startTime, jobs[curr][1])
            max_profit = max(findMaxProfit(curr + 1), jobs[curr][2] + findMaxProfit(next_idx))
            memo[curr] = max_profit
            return memo[curr]
        
        n = len(startTime)
        jobs = []
        for s, e, p in zip(startTime, endTime, profit):
            jobs.append([s, e, p])
        jobs.sort()
        startTime.sort()
        memo = [-1] * n
        return findMaxProfit(0)